### DISCUSSION TITLE
Problem connecting to optitrack mocap

### DISCUSSION THREAD

**Discussion by manvibharat** [USC-ACTLab/crazyswarm#335] [Q&A]
I have tried https://github.com/whoenig/NatNetSDKCrossplatform and ./mcr works well and I am getting the data there. But when I run mocap_helper.launch with the following contents

---
      # tracking
      motion_capture_type: "optitrack" # one of vicon,optitrack,qualisys,vrpn
      object_tracking_type: "motionCapture" # one of motionCapture,libobjecttracker
      # vicon_host_name: "vicon"
      optitrack_host_name: "optitrack"
      #optitrack_local_ip: "192.168.43.238"
      #optitrack_server_ip: "192.168.43.73"
      # qualisys_host_name: "10.0.5.219"
      # qualisys_base_port: 22222
      # vrpn_host_name: "vicon"
---
 -> it gives the following error ---

---

started roslaunch server http://rosmelodic:40259/

PARAMETERS
 * /mocap_helper/motion_capture_type: optitrack
 * /mocap_helper/object_tracking_type: motionCapture
 * /mocap_helper/optitrack_host_name: optitrack
 * /rosdistro: melodic
 * /rosversion: 1.14.10

NODES
  /
    mocap_helper (crazyswarm/mocap_helper)
ROS_MASTER_URI=http://localhost:11311
process[mocap_helper-1]: started with pid [23540]
frame 0:
[mocap_helper-1] process has died [pid 23540, exit code -11, cmd /home/rosmelodic/crazyswarm/ros_ws/devel/lib/crazyswarm/mocap_helper __name:=mocap_helper __log:=/home/rosmelodic/.ros/log/03efa038-7851-11eb-88d8-c4041567d7d3/mocap_helper-1.log].
log file: /home/rosmelodic/.ros/log/03efa038-7851-11eb-88d8-c4041567d7d3/mocap_helper-1*.log
all processes on machine have died, roslaunch will exit
shutting down processing monitor...
... shutting down processing monitor complete
done

---
However when I use methods described in some of the issues like using local ip and server ip of optitrack; the terminal stays blank no points are printed.

The ./mcr output is as follows

---

Begin Packet
Message ID : 7
Byte count : 96
Frame # : 11627530
Marker Set Count : 0
Unidentified Marker Count : 0
Rigid Body Count : 1
ID : 1
pos: [1.60,-1.17,-0.00]
ori: [-0.00,-0.00,-0.17,-0.98]
software latency : 0.000
Timestamp : 0.000
Mid-exposure timestamp : 4530722105326567424
Camera data received timestamp : 0
Transmit timestamp : 4682726790983896269
End Packet

---
Also I downloaded NatNetSDK from official source and placed in external dependencies but no luck with that.

So any solutions possible?

**Comment by whoenig** [comment]
How did you run `./mcr`, since that requires an IP address to be provided? Local and server IP are not used anymore. The only parameter you need to set is optitrack_host_name, see https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/mocap_helper.cpp#L51.

**Comment by whoenig** [comment]
You don't need to download NatNetSDK manually. Make sure you use the version we tested with, that is included as a git submodule. The crash you are observing is likely related to some change in the packet format.

**Comment by manvibharat** [comment]
> How did you run `./mcr`, since that requires an IP address to be provided? Local and server IP are not used anymore. The only parameter you need to set is optitrack_host_name, see https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/mocap_helper.cpp#L51.

Yeah I used the ip address of the host which I saved in /etc/hosts

**Comment by whoenig** [comment]
You should use the same argument in mocap_helper; just because it is called hostname, doesn't mean that it can't take an IP address.

**Comment by manvibharat** [comment]
@whoenig 
Also if it helps I am using Optitrack 1.10.3 version. Do you think it is the problem?. Also if possible can you suggest which file I should look to change for this to change packet format. 
Thanks

**Comment by whoenig** [comment]
We only tested with Motive 2.X. For the packet format, it would be best to find out where the crash occurs (see http://wiki.ros.org/roslaunch/Tutorials/Roslaunch%20Nodes%20in%20Valgrind%20or%20GDB on how to run a ROS node with a debugger). Then you can compare the packet decoding in that part of the code with the one that is used in the NatNetSDK example.

**Comment by lkyfly0765** [comment]
It seems that your launch file is not correctly configured.
It should be:
```
optitrack_local_ip: "192.168.43.238"
optitrack_server_ip: "192.168.43.73"
```
NOT  with '#'
Also, I strongly suggest that you should upgrade your Motive software to 2.x version.

For the packet format concerns by @whoenig , please make sure that in Motive-Datastreaming panel, only the rigid body is select. Do not select anything contains 'makers' if you use the motioncapture mode, or there will be some packet format errors.

**Reply by lkyfly0765** [reply]
The ./mcr outputs the correct data means that the network connection between ROS and Motive is good.
